datasource/src/main/scala/quasar/physical/mongo/Mongo.scala

```diff
@@ -69,46 +72,73 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
 
     (for {
-      obsQ <- Stream.eval(Queue.boundedNoneTerminated[F, Either[Throwable, A]](queueSize))
+      obsQ <- Stream.eval(Queue.boundedNoneTerminated[F, Either[Throwable, A]](queueSize.toInt))
       subVar <- Stream.eval(MVar[F].empty[Subscription])
       _ <- enqueueObservable(obsQ, subVar)
-      res <- obsQ.dequeueChunk(queueSize).chunks.flatMap( c =>
+      res <- obsQ.dequeue.chunks.flatMap( c =>
         Stream.evalUnChunk(subVar.read.map(_.request(c.size.toLong)).as(c))
       ).rethrow.onFinalize(subVar.take.flatMap(unsubscribe))
     } yield res)
   }
 
+
+
   def databases: Stream[F, Database] =
-    observableAsStream(client.listDatabaseNames).map(Database(_))
+    observableAsStream(client.listDatabaseNames, DefaultQueueSize).map(Database(_))
 
   def databaseExists(database: Database): Stream[F, Boolean] =
     databases.exists(_ === database)
 
   def collections(database: Database): Stream[F, Collection] = for {
     dbExists <- databaseExists(database)
     res <- if (!dbExists) { Stream.empty } else {
-      observableAsStream(client.getDatabase(database.name).listCollectionNames()).map(Collection(database, _))
+      observableAsStream(client.getDatabase(database.name).listCollectionNames(), DefaultQueueSize).map(Collection(database, _))
     }
   } yield res
 
   def collectionExists(collection: Collection): Stream[F, Boolean] =
     collections(collection.database).exists(_ === collection)
 
   def findAll(collection: Collection): F[Stream[F, BsonValue]] = {
-    def getCollection(mongo: MongoClient) =
-      mongo.getDatabase(collection.database.name).getCollection(collection.name)
-
     collectionExists(collection).compile.last.map(_ getOrElse false)
       .flatMap(exists =>
         if (exists) {
-          F.delay(observableAsStream(getCollection(client).find[BsonValue]()))
+          getQueueSize(collection).map { (qs: Long) =>
+            observableAsStream(getCollection(collection).find[BsonValue](), qs)
+          }
         } else {
           MonadResourceErr.raiseError(ResourceError.pathNotFound(collection.resourcePath))
         })
   }
+
+  def getQueueSize(collection: Collection): F[Long] = {
+    val getStats: F[Document] = F.suspend(singleObservableAsF[F, Document](
+      client.getDatabase(collection.database.name).runCommand[Document](Document(
+        "collStats" -> collection.name,
+        "scale" -> 1
+      ))
+    ))
+
+    val max = if (maxMemory > MaxBsonBatch) MaxBsonBatch else maxMemory.toLong
+
+    getStats map { doc =>
+      val avgObjSize = doc.get("avgObjSize").fold(0)(_.asInt32.getValue)
+      val res = max / avgObjSize.toLong
+      if (res < 1L) 1L else if (res < MaxQueueSize) res else MaxQueueSize
+    }
+  }
+
+  private def getCollection(collection: Collection): MongoCollection[Document] =
+    client.getDatabase(collection.database.name).getCollection(collection.name)
+
 }
 
 object Mongo {
+  val DefaultBsonBatch: Int = 1024 * 128
```

Comment:
With absolutely zero evidence to support this suggestion, I think defaulting to 2MB would be a bit nicer for throughput as 128k seems pretty small.